File: Content/panorama/scripts/custom_game/elements/elements.js
function convertElementNameToNumber(element_name)
{
	var element_number = 0
	if (element_name == "nature"){
		element_number = 16
	}else if(element_name == "cosmic"){
		element_number = 8
	}else if(element_name == "lightning"){
		element_number = 4
	}else if(element_name == "fire"){
		element_number = 2
	}
	return element_number
}

function convertElementNumberToName(element_number)
{
	var element_name = ""
	if (element_number == 1){
		element_name = "normal"
	}else if(element_number == 2){
		element_name = "fire"
	}else if(element_number == 3){
		element_name = "earth"
	}else if(element_number == 4){
		element_name = "lightning"
	}else if(element_number == 5){
		element_name = "poison"
	}else if(element_number == 6){
		element_name = "time"
	}else if(element_number == 7){
		element_name = "holy"
	}else if(element_number == 8){
		element_name = "cosmic"
	}else if(element_number == 9){
		element_name = "ice"
	}else if(element_number == 10){
		element_name = "arcane"
	}else if(element_number == 11){
		element_name = "shadow"
	}else if(element_number == 12){
		element_name = "wind"
	}else if(element_number == 13){
		element_name = "ghost"
	}else if(element_number == 14){
		element_name = "water"
	}else if(element_number == 15){
		element_name = "demon"
	}else if(element_number == 16){
		element_name = "nature"
	}else if(element_number == 17){
		element_name = "undead"
	}else if(element_number == 18){
		element_name = "dragon"
	}
	return element_name
}

function get_element_color_by_index(element_index){
	var color = "#FFFFFF"
	if (element_index == 1){
		color = "#DDDDDD"
	}else if(element_index == 2){
		color = "#EF4126"
	}else if(element_index == 3){
		color = "#AF843D"
	}else if(element_index == 4){
		color = "#5CCDF9"
	}else if(element_index == 5){
		color = "#37DD3D"
	}else if(element_index == 6){
		color = "#B5FFB7"
	}else if(element_index == 7){
		color = "#F6FFB5"
	}else if(element_index == 8){
		color = "#C25DFC"
	}else if(element_index == 9){
		color = "#87D9FF"
	}else if(element_index == 10){
		color = "#E1A2E8"
	}else if(element_index == 11){
		color = "#7F4F84"
	}else if(element_index == 12){
		color = "#7AE2A7"
	}else if(element_index == 13){
		color = "#9ACCD1"
	}else if(element_index == 14){
		color = "#3894FF"
	}else if(element_index == 15){
		color = "#5B648C"
	}else if(element_index == 16){
		color = "#69BC71"
	}else if(element_index == 17){
		color = "#5C776E"
	}
	return color
}

File: Game/scripts/vscripts/elements.lua
if Elements == nil then
  Elements = class({})
end

RPC_ELEMENT_COUNT = 18

RPC_ELEMENT_NONE = -1
RPC_ELEMENT_NORMAL = 1
RPC_ELEMENT_FIRE = 2
RPC_ELEMENT_EARTH = 3
RPC_ELEMENT_LIGHTNING = 4
RPC_ELEMENT_POISON = 5
RPC_ELEMENT_TIME = 6
RPC_ELEMENT_HOLY = 7
RPC_ELEMENT_COSMOS = 8
RPC_ELEMENT_ICE = 9
RPC_ELEMENT_ARCANE = 10
RPC_ELEMENT_SHADOW = 11
RPC_ELEMENT_WIND = 12
RPC_ELEMENT_GHOST = 13
RPC_ELEMENT_WATER = 14
RPC_ELEMENT_DEMON = 15
RPC_ELEMENT_NATURE = 16
RPC_ELEMENT_UNDEAD = 17
RPC_ELEMENT_DRAGON = 18

RPC_ELEMENT_NORMAL_COLOR = "#DDDDDD"
RPC_ELEMENT_FIRE_COLOR = "#EF4126"
RPC_ELEMENT_EARTH_COLOR = "#AF843D"
RPC_ELEMENT_LIGHTNING_COLOR = "#5CCDF9"
RPC_ELEMENT_POISON_COLOR = "#37DD3D"
RPC_ELEMENT_TIME_COLOR = "#B5FFB7"
RPC_ELEMENT_HOLY_COLOR = "#F6FFB5"
RPC_ELEMENT_COSMOS_COLOR = "#C25DFC"
RPC_ELEMENT_ICE_COLOR = "#87D9FF"
RPC_ELEMENT_ARCANE_COLOR = "#E1A2E8"
RPC_ELEMENT_SHADOW_COLOR = "#7F4F84"
RPC_ELEMENT_WIND_COLOR = "#7AE2A7"
RPC_ELEMENT_GHOST_COLOR = "#9ACCD1"
RPC_ELEMENT_WATER_COLOR = "#3894FF"
RPC_ELEMENT_DEMON_COLOR = "#5B648C"
RPC_ELEMENT_NATURE_COLOR = "#69BC71"
RPC_ELEMENT_UNDEAD_COLOR = "#5C776E"
RPC_ELEMENT_DRAGON_COLOR = "#3289C7"

function Elements:GetElementIndexByString(element_name)
	local element_number = 0
	if element_name == "normal" then
		element_number = RPC_ELEMENT_NORMAL
	elseif element_name == "fire" then
		element_number = RPC_ELEMENT_FIRE
	elseif element_name == "earth" then
		element_number = RPC_ELEMENT_EARTH
	elseif element_name == "lightning" then
		element_number = RPC_ELEMENT_LIGHTNING
	elseif element_name == "poison" then
		element_number = RPC_ELEMENT_POISON
	elseif element_name == "time" then
		element_number = RPC_ELEMENT_TIME
	elseif element_name == "holy" then
		element_number = RPC_ELEMENT_HOLY
	elseif element_name == "cosmic" then
		element_number = RPC_ELEMENT_COSMOS
	elseif element_name == "ice" then
		element_number = RPC_ELEMENT_ICE
	elseif element_name == "arcane" then
		element_number = RPC_ELEMENT_ARCANE
	elseif element_name == "shadow" then
		element_number = RPC_ELEMENT_SHADOW
	elseif element_name == "wind" then
		element_number = RPC_ELEMENT_WIND
	elseif element_name == "ghost" then
		element_number = RPC_ELEMENT_GHOST
	elseif element_name == "water" then
		element_number = RPC_ELEMENT_WATER
	elseif element_name == "demon" then
		element_number = RPC_ELEMENT_DEMON
	elseif element_name == "nature" then
		element_number = RPC_ELEMENT_NATURE
	elseif element_name == "undead" then
		element_number = RPC_ELEMENT_UNDEAD
	elseif element_name == "dragon" then
		element_number = RPC_ELEMENT_DRAGON
	end
	return element_number
end

function Elements:GetElementNameAndColorByCode(elementCode)
	local name = ""
	local color = "#000000"
	if elementCode == RPC_ELEMENT_NORMAL then
		name = "normal"
		color = RPC_ELEMENT_NORMAL_COLOR
	elseif elementCode == RPC_ELEMENT_FIRE then
		name = "fire"
		color = RPC_ELEMENT_FIRE_COLOR
	elseif elementCode == RPC_ELEMENT_EARTH then
		name = "earth"
		color = RPC_ELEMENT_EARTH_COLOR
	elseif elementCode == RPC_ELEMENT_LIGHTNING then
		name = "lightning"
		color = RPC_ELEMENT_LIGHTNING_COLOR
	elseif elementCode == RPC_ELEMENT_POISON then
		name = "poison"
		color = RPC_ELEMENT_POISON_COLOR
	elseif elementCode == RPC_ELEMENT_TIME then
		name = "time"
		color = RPC_ELEMENT_TIME_COLOR
	elseif elementCode == RPC_ELEMENT_HOLY then
		name = "holy"
		color = RPC_ELEMENT_HOLY_COLOR
	elseif elementCode == RPC_ELEMENT_COSMOS then
		name = "cosmos"
		color = RPC_ELEMENT_COSMOS_COLOR
	elseif elementCode == RPC_ELEMENT_ICE then
		name = "ice"
		color = RPC_ELEMENT_ICE_COLOR
	elseif elementCode == RPC_ELEMENT_ARCANE then
		name = "arcane"
		color = RPC_ELEMENT_ARCANE_COLOR
	elseif elementCode == RPC_ELEMENT_SHADOW then
		name = "shadow"
		color = RPC_ELEMENT_SHADOW_COLOR
	elseif elementCode == RPC_ELEMENT_WIND then
		name = "wind"
		color = RPC_ELEMENT_WIND_COLOR
	elseif elementCode == RPC_ELEMENT_GHOST then
		name = "ghost"
		color = RPC_ELEMENT_GHOST_COLOR
	elseif elementCode == RPC_ELEMENT_WATER then
		name = "water"
		color = RPC_ELEMENT_WATER_COLOR
	elseif elementCode == RPC_ELEMENT_DEMON then
		name = "demon"
		color = RPC_ELEMENT_DEMON_COLOR
	elseif elementCode == RPC_ELEMENT_NATURE then
		name = "nature"
		color = RPC_ELEMENT_NATURE_COLOR
	elseif elementCode == RPC_ELEMENT_UNDEAD then
		name = "undead"
		color = RPC_ELEMENT_UNDEAD_COLOR
	elseif elementCode == RPC_ELEMENT_DRAGON then
		name = "dragon"
		color = RPC_ELEMENT_DRAGON_COLOR
	end
	return name, color
end

function Elements:hex2rgb(hex)
    hex = hex:gsub("#","")
    return tonumber("0x"..hex:sub(1,2)), tonumber("0x"..hex:sub(3,4)), tonumber("0x"..hex:sub(5,6))
end

function Elements:RollElementAttribute(item, element_code, rollMult, minRoll, maxRoll, attributeSlot)
	local luck = RandomInt(0,110)
	local maxFactor = RPCItems:GetMaxFactor()
	local value = 0
	local prefixLevel = 1
	local maxRoll = math.ceil(rollMult*1.5)
	local name, color = Elements:GetElementNameAndColorByCode(element_code)
	value, prefixLevel = RPCItems:RollAttribute(100, minRoll, maxRoll, 0, 0, item.rarity, false, maxFactor*rollMult)
	value = math.floor(value)
	print("[Elements:RollElementAttribute] "..name)
	if attributeSlot == 1 then
		item.newItemTable.property1name = name
		RPCItems:SetPropertyValues(item, value, "#rpc_item_element"..element_code, color,  1)
	elseif attributeSlot == 2 then
		item.newItemTable.property2name = name
		RPCItems:SetPropertyValues(item, value, "#rpc_item_element"..element_code, color,  2)
	elseif attributeSlot == 3 then
		item.newItemTable.property3name = name
		RPCItems:SetPropertyValues(item, value, "#rpc_item_element"..element_code, color,  3)
	elseif attributeSlot == 4 then
		item.newItemTable.property4name = name
		RPCItems:SetPropertyValues(item, value, "#rpc_item_element"..element_code, color,  4)
	end
end